File: solutions/problem_1239_maximum_length_of_a_concatenated_string_with_unique_characters/solution.cpp
/*
 * Problem 1239: Maximum Length of a Concatenated String with Unique Characters
 * ==========================================================================
 * Difficulty: Medium
 * Tags: Array, String, Backtracking, Bit Manipulation
 * Pattern: Backtracking
 *
 * Time Complexity:  O(k^n) or O(n!)
 * Space Complexity: O(n)
 */

#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int maxLength(vector<string>& arr) {
        // Backtracking - O(2^n) or O(n!) time
        vector<vector<int>> result;
        vector<int> path;
        function<void(int)> backtrack = [&](int start) {
            result.push_back(path);
            for (int i = start; i < (int)arr.size(); i++) {
                path.push_back(arr[i]);
                backtrack(i + 1);
                path.pop_back();
            }
        };
        backtrack(0);
        return result;
    }
};

File: solutions/problem_1239_maximum_length_of_a_concatenated_string_with_unique_characters/solution.py
"""
Problem 1239: Maximum Length of a Concatenated String with Unique Characters
==========================================================================
Difficulty: Medium
Tags: Array, String, Backtracking, Bit Manipulation
Pattern: Backtracking

Time Complexity:  O(k^n) or O(n!)
Space Complexity: O(n)
"""

class Solution:
    def maxLength(self, arr: List[str]) -> int:
        # Backtracking - O(2^n) or O(n!) time
        result = []
        
        def backtrack(path, start):
            result.append(path[:])
            for i in range(start, len(arr)):
                path.append(arr[i])
                backtrack(path, i + 1)
                path.pop()
        
        backtrack([], 0)
        return result
